### PR TITLE
Add support for `zstd` compression

### PR DESCRIPTION
**Pull request by gabriel-tincu**
The ZSTD relevant code, in a separate PR, @jeffwidman

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/2021)

### CODE REVIEW

File: test/test_producer.py [hunk 1]
```diff
@@ -34,10 +34,15 @@ def test_end_to_end(kafka_broker, compression):
         elif platform.python_implementation() == 'PyPy':
             return
 
+    if compression == 'zstd' and env_kafka_version() < (2, 1, 0):
+        return
+    env_version = env_kafka_version()
+    api_version = env_version if env_version >= (2, 1, 0) else None
```
Comment (jeffwidman) [inline test/test_producer.py]: What is the purpose of adding this? I'm not saying it's incorrect, I simply don't understand why the test didn't previously use this and then now needs it...

File: test/test_producer.py [hunk 2]
```diff
@@ -34,10 +34,15 @@ def test_end_to_end(kafka_broker, compression):
         elif platform.python_implementation() == 'PyPy':
             return
 
+    if compression == 'zstd' and env_kafka_version() < (2, 1, 0):
+        return
```
Comment (jeffwidman) [inline test/test_producer.py]: This should be using a `pytest` skip... http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions
Could be done inside the test function, but IMO the decorator pattern is more readable when skimming the tests... since decorators fire as the inner ones first, I'd use the following order:
```
@pytest.mark.skipif(compression == 'zstd' and env_kafka_version() < (2, 1, 0), reason="zstd support requires Kafka 2.1 (KIP-110)")
@pytest.mark.skipif(not env_kafka_version(), reason="No KAFKA_VERSION set")
@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4', 'zstd'])
```

File: test/test_producer.py [hunk 3]
```diff
@@ -81,11 +86,16 @@ def test_kafka_producer_gc_cleanup():
 
 
 @pytest.mark.skipif(not env_kafka_version(), reason="No KAFKA_VERSION set")
-@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4'])
+@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4', 'zstd'])
 def test_kafka_producer_proper_record_metadata(kafka_broker, compression):
+    if compression == 'zstd' and env_kafka_version() < (2, 1, 0):
+        return
```
Comment (jeffwidman) [inline test/test_producer.py]: see note above about using a decorator instead to actually skip the test rather than considering it as passed

**Comment by jeffwidman** [comment]
Also, in https://kafka-python.readthedocs.io/en/master/#compression can you mention support for `zstd`?

A clean way to do it would be to make it a bullet pointed list...
> kafka-python supports the following compression formats:
> * gzip  - support built-in 
> * lz4 - requires installing `python-lz4`
> * snappy - requires installing `python-snappy` (which requires the `snappy` `C` library)
> * zstd - requires installing `zstandard`

File: test/test_producer.py (hunk 2 above)
Comment (gabriel-tincu) [reply test/test_producer.py]: > This should be using a `pytest` skip... http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions
> Could be done inside the test function, but IMO the decorator pattern is more readable when skimming the tests... since decorators fire as the inner ones first, I'd use the following order:
> 
> ```
> @pytest.mark.skipif(compression == 'zstd' and env_kafka_version() < (2, 1, 0), reason="zstd support requires Kafka 2.1 (KIP-110)")
> @pytest.mark.skipif(not env_kafka_version(), reason="No KAFKA_VERSION set")
> @pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4', 'zstd'])
> ```

The compression param is injected by one of the decorators though, so that notation will not work, but i'll give the doc a read and figure out if i can manipulate them after the `parametrize` decorator is in place

File: test/test_producer.py (hunk 1 above)
Comment (gabriel-tincu) [reply test/test_producer.py]: the point was to force an api version injection only if the env var defined api si larger than 2.1, else let the connector handle it by querying the broker. Now that you mention it, i don't see how it could actually help, but i do remember some weird behaviour on a combination of detected broker version and actual broker version, since they don't always seem to match. I'll leave that part out and come back to it if i have a clear scenario when X does not play nicely with Y

File: test/test_producer.py (hunk 3 above)
Comment (gabriel-tincu) [reply test/test_producer.py]: Cannot figure a way to make this work without the conditional , since there is no (apparent to me) way of accessing the parameter injected by one decorator into the other one.

File: test/test_producer.py (hunk 2 above)
Comment (jeffwidman) [reply test/test_producer.py]: oh interesting... I was expecting it to wrap the function in a way that made it possible to introspect it... essentially each decorator gives a wrapped function to the one above it.

Instead of a decorator, it's also possible to use `pytest.skip("reason")` function within a test... use that in place of your `return` statement. More details in http://doc.pytest.org/en/latest/skipping.html#skipping-test-functions

**Comment by jeffwidman** [comment]
Now that https://github.com/dpkp/kafka-python/pull/2020 is merged, can you please rebase this on top of that?

**Comment by gabriel-tincu** [comment]
> Now that #2020 is merged, can you please rebase this on top of that?

done, though it seems github diff algorithm or needs to refresh master the branch tip

**Comment by jeffwidman** [comment]
The diff is appearing fine for me.

Typically what you'd do in this case is `git rebase -i origin/master` and then squash all these commits into a single one...or just a couple of logically separated ones... ie, the fixup commits get merged together.

File: kafka/codec.py [hunk 4]
```diff
@@ -299,3 +309,18 @@ def lz4_decode_old_kafka(payload):
...
+    return zstd.ZstdCompressor().compress(payload)
+
+
+def zstd_decode(payload):
+    if not zstd:
+        raise NotImplementedError("Zstd codec is not available")
+    try:
+        return zstd.ZstdDecompressor().decompress(payload)
+    except zstd.ZstdError:
+        return zstd.ZstdDecompressor().decompress(payload, max_output_size=ZSTD_MAX_OUTPUT_SIZE)
```
Comment (jeffwidman) [inline kafka/codec.py]: Reading the docs on [`ZstdError`](https://github.com/indygreg/python-zstandard#simple-api-1) it looks like this can still cause problems if the output is larger than `ZSTD_MAX_OUTPUT_SIZE`... 

It looks like the value of `ZSTD_MAX_OUTPUT_SIZE` is currently set to `1024 ** 3`... Why did you pick that particular value?

I realize this is a fallback, with the expectation that in the normal case the zstd header will contain the frame size, so I think it's okay to set a fallback that works for the majority of cases but may not work in all cases...

**Comment by jeffwidman** [comment]
Actually strike that, in this diff I'm suddenly seeing the `log_start_offset` stuff that has already been merged...

Are you sure you pulled latest `master` and then rebased on top of that?

File: test/test_producer.py [hunk 5]
```diff
@@ -23,21 +23,22 @@ def test_buffer_pool():
...
-@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4'])
+@pytest.mark.parametrize("compression", [None, 'gzip', 'snappy', 'lz4', 'zstd'])
 def test_end_to_end(kafka_broker, compression):
-
     if compression == 'lz4':
-        # LZ4 requires 0.8.2
         if env_kafka_version() < (0, 8, 2):
-            return
-        # python-lz4 crashes on older versions of pypy
+            pytest.skip('LZ4 requires 0.8.2')
```
Comment (jeffwidman) [inline test/test_producer.py]: thanks for this cleanup!

File: test/test_producer.py [hunk 6]
```diff
@@ -23,21 +23,22 @@ def test_buffer_pool():
...
+            pytest.skip('python-lz4 crashes on older versions of pypy')
+
+    if compression == 'zstd' and env_kafka_version() < (2, 1, 0):
+        pytest.skip('zstd requires kafka 2.1.0 or newer')
 
     connect_str = ':'.join([kafka_broker.host, str(kafka_broker.port)])
     producer = KafkaProducer(bootstrap_servers=connect_str,
                              retries=5,
                              max_block_ms=30000,
+                             api_version=env_kafka_version(),
```
Comment (jeffwidman) [inline test/test_producer.py]: Why is this needed when it wasn't needed previously?

**Comment by gabriel-tincu** [comment]
> Actually strike that, in this diff I'm suddenly seeing the `log_start_offset` stuff that has already been merged...
> 
> Are you sure you pulled latest `master` and then rebased on top of that?

i rebased on top of the merged branch, due to both PR's being made from a forked version of the repo. That is to say, that repo's master is not in sync with this repo's master, for various reasons. 
That is why i linked the compare link, which compares the tips of both branches (or rather tip-1 for the source branch, as i made a commit just a few minutes ago), to illustrate that using compare, the diff appears as it should

**Comment by jeffwidman** [comment]
> i rebased on top of the merged branch, due to both PR's being made from a forked version of the repo. That is to say, that repo's master is not in sync with this repo's master, for various reasons.
That is why i linked the compare link, which compares the tips of both branches (or rather tip-1 for the source branch, as i made a commit just a few minutes ago), to illustrate that using compare, the diff appears as it should

Sorry, I wasn't clear... you don't want to rebase on top of the merged PR branch, you want to rebase on top of `master` of this repo, which contains the merged branch...

We'll need to see the diff so we can see _exactly_ what is getting merged.

File: kafka/codec.py (hunk 4 above)
Comment (gabriel-tincu) [reply kafka/codec.py]: I actually made this fail using the java console producer, which does not properly complete the message frame with size information. I used this line https://github.com/facebook/zstd/blob/dev/doc/educational_decoder/harness.c#L90 as inspiration, and while i do realise that value does not come into play in any shape inside the core algorithm, it was better than nothing

**Comment by gabriel-tincu** [comment]
> > i rebased on top of the merged branch, due to both PR's being made from a forked version of the repo. That is to say, that repo's master is not in sync with this repo's master, for various reasons.
> > That is why i linked the compare link, which compares the tips of both branches (or rather tip-1 for the source branch, as i made a commit just a few minutes ago), to illustrate that using compare, the diff appears as it should
> 
> Sorry, I wasn't clear... you don't want to rebase on top of the merged PR branch, you want to rebase on top of `master` of this repo, which contains the merged branch...
> 
> We'll need to see the diff so we can see _exactly_ what is getting merged.

For future reference, that would imply updating the origin first, then doing the actual rebase, or just adding a new remote, correct? I'll try to open a new PR with an identical but differently named branch

File: test/test_producer.py (hunk 6 above)
Comment (gabriel-tincu) [reply test/test_producer.py]: auto discovered version would always come as 1.0.0 or similar. I think there is a PR now that actually deals with that  and updates the `conn.py::check_version` code, but i needed to force i higher version in order for the broker to not reject the requests

File: kafka/codec.py (hunk 4 above)
Comment (gabriel-tincu) [reply kafka/codec.py]: though in retrospect this will probably end up causing a lot of OOM errors in a traffic intensive scenario, since it tries to allocate 1G or memory per message with frame information missing. Ideally this would be taken from the broker configuration `"message.max.bytes"`, but i see no way of passing that in

Comment (jeffwidman) [reply kafka/codec.py]: Agreed that `message.max.bytes` would be ideal... for now, why don't we just set it to 1 MB, since that's the default for `message.max.bytes`?

I expect that won't be perfect, but it'll be a start and let us land this and then see what further suggestions/feedback arise from others... some of this ultimately comes back to it'd be better if we could let compression be a passed in class so that users can tweak their own settings external to the client. But that's a much bigger change far beyond the scope of this PR.

File: test/test_producer.py (hunk 6 above)
Comment (jeffwidman) [reply test/test_producer.py]: Gotcha, that implies that for end users who are expecting this to just work, it won't work until the check version code is fixed. Hmm... I'd kinda rather see that merged first, I don't recall how much work that is, I'd have to check...

**Comment by jeffwidman** [comment]
Can you just merge `master` into this one? It should fix the diff w/o needing to rebase or anything...

File: kafka/codec.py (hunk 4 above)
Comment (gabriel-tincu) [reply kafka/codec.py]: done

**Comment by gabriel-tincu** [comment]
> Can you just pull latest `master` from this repo and then merge that `master` into this branch? It should fix the diff w/o needing to rebase or anything...
> 
> And then I'll cleanup the history when I squash merge. If this truly doesn't work, then it's a weird github bug, but I expect it should work since we often do that workflow at my day job with no issues.

That seems to have worked

**Comment by jeffwidman** [comment]
IMO this is mostly ready to go, except for https://github.com/dpkp/kafka-python/pull/2021#discussion_r398247782... fixing the root issue of that is outside the scope of this particular PR, but in the meantime I'd rather not add that workaround hack.

It looks like #2024 takes a stab at fixing the root cause, let's see if we can land that and then land this...

**Comment by Green-Angry-Bird** [comment]
Our team is looking for exactly this. When can we merge to master. :D

**Comment by tvoinarovskyi** [comment]
@gabriel-tincu Sorry, but could you rebase this over master now? Thanks =)

**Comment by gabriel-tincu** [comment]
> @gabriel-tincu Sorry, but could you rebase this over master now? Thanks =)

@tvoinarovskyi Done. I noticed that tox fails specifically for python3.7, but with an error related to tox itself rather than anything related to this PR

File: test/test_producer.py (hunk 6 above)
Comment (gabriel-tincu) [reply test/test_producer.py]: fixed

**Comment by Green-Angry-Bird** [comment]
👍 I can't wait for my consumer to not crash on zstd encoded messages.

**Comment by Green-Angry-Bird** [comment]
Is there any update on this?  Our team is looking forward to using this library with our zstd encoded topics.

**Comment by Green-Angry-Bird** [comment]
Is this ready for merge?

**Comment by brokenjacobs** [comment]
ping?

**Comment by Green-Angry-Bird** [comment]
I'm a little worried kafka-python is no longer maintained.

For anyone else experiencing this issue; I've created a workaround. I created this workaround in the hopes that kafka-python would be supporting zstd in the near future. While not ideal, this solution pipes the data from the zstd topic into a new topic which has been configured with gzip compression. It requires command-line access on the broker and can be monitored in the Kafka Manager via group ID.

Step 1: Create a new topic and be sure to set compression.type=gzip 
`kafka-topics --create --bootstrap-server $BROKERS --replication-factor $REPLICAS --partitions $PARTITIONS --config compression.type=gzip --config retention.ms=$RETENTION_MS --config retention.bytes=$RETENTION_BYTES --topic $TOPIC-gzip`

Step 2: Create / Reset a consumer group (to track the mirror)
`kafka-consumer-groups --bootstrap-server $BROKERS --group $GROUP --topic $TOPIC --reset-offsets --to-earliest --execute`

Step 3: Begin the mirror threads
`kafka-console-consumer --bootstrap-server $BROKERS --group $GROUP --topic $TOPIC | kafka-console-producer --broker-list $BROKERS  --compression-codec gzip --topic $TOPIC-gzip > $TOPIC.log 2> $TOPIC.err`

I hope this helps someone else.

**Comment by gabriel-tincu** [comment]
> I'm a little worried kafka-python is no longer maintained.
> 
> For anyone else experiencing this issue; I've created a workaround. I created this workaround in the hopes that kafka-python would be supporting zstd in the near future. While not ideal, this solution pipes the data from the zstd topic into a new topic which has been configured with gzip compression. It requires command-line access on the broker and can be monitored in the Kafka Manager via group ID.
> 
> Step 1: Create a new topic and be sure to set compression.type=gzip
> `kafka-topics --create --bootstrap-server $BROKERS --replication-factor $REPLICAS --partitions $PARTITIONS --config compression.type=gzip --config retention.ms=$RETENTION_MS --config retention.bytes=$RETENTION_BYTES --topic $TOPIC-gzip`
> 
> Step 2: Create / Reset a consumer group (to track the mirror)
> `kafka-consumer-groups --bootstrap-server $BROKERS --group $GROUP --topic $TOPIC --reset-offsets --to-earliest --execute`
> 
> Step 3: Begin the mirror threads
> `kafka-console-consumer --bootstrap-server $BROKERS --group $GROUP --topic $TOPIC | kafka-console-producer --broker-list $BROKERS --compression-codec gzip --topic $TOPIC-gzip > $TOPIC.log 2> $TOPIC.err`
> 
> I hope this helps someone else.

@Green-Angry-Bird you could just use the source branch HEAD of this PR in your requirements file (assuming that's how you handle dependencies) until this gets merged. Taking the current global situation into account i think some delay in response on an open source project is perfectly understandable

**Comment by Green-Angry-Bird** [comment]
> @Green-Angry-Bird you could just use the source branch HEAD of this PR in your requirements file (assuming that's how you handle dependencies) until this gets merged. Taking the current global situation into account i think some delay in response on an open source project is perfectly understandable

Forgive my ignorance @gabriel-tincu but how would I do that?

I've tried these but they all give different errors:
```-e git://github.com/aiven/kafka-python/tree/gabi-kafka-client-zstd-codec#egg=kafka-python
-e git://github.com/dpkp/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python
-e git://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python```

**Comment by gabriel-tincu** [comment]
> > @Green-Angry-Bird you could just use the source branch HEAD of this PR in your requirements file (assuming that's how you handle dependencies) until this gets merged. Taking the current global situation into account i think some delay in response on an open source project is perfectly understandable
> 
> Forgive my ignorance @gabriel-tincu but how would I do that?
> 
> I've tried these but they all give different errors:
> 
> ```
> -e git://github.com/aiven/kafka-python/tree/gabi-kafka-client-zstd-codec#egg=kafka-python
> -e git://github.com/dpkp/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python
> -e git://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python
> ```

What error does the third option throw? Also, have you tried using it without the `#egg=kafka-python` at the end? 
`pip install git+https://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec` seems to work in a brand new virtualenv without any errors for me

**Comment by Green-Angry-Bird** [comment]
> > > @Green-Angry-Bird you could just use the source branch HEAD of this PR in your requirements file (assuming that's how you handle dependencies) until this gets merged. Taking the current global situation into account i think some delay in response on an open source project is perfectly understandable
> > 
> > 
> > Forgive my ignorance @gabriel-tincu but how would I do that?
> > I've tried these but they all give different errors:
> > ```
> > -e git://github.com/aiven/kafka-python/tree/gabi-kafka-client-zstd-codec#egg=kafka-python
> > -e git://github.com/dpkp/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python
> > -e git://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python
> > ```
> 
> What error does the third option throw? Also, have you tried using it without the `#egg=kafka-python` at the end?
> `pip install git+https://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec` seems to work in a brand new virtualenv without any errors for me

The error was
```
$ pip3 install -r requirements.txt 
Obtaining kafka-python from git+git://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python (from -r requirements.txt (line 2))
  WARNING: git clone in ./src/kafka-python exists with URL git://github.com/dpkp/kafka-python
  WARNING: The plan is to install the git repository git://github.com/aiven/kafka-python
What to do?  (s)witch, (i)gnore, (w)ipe, (b)ackup s
  Switching clone ./src/kafka-python to git://github.com/aiven/kafka-python (to revision gabi-kafka-client-zstd-codec)
  Running command git config remote.origin.url git://github.com/aiven/kafka-python
  Running command git checkout -q gabi-kafka-client-zstd-codec
  error: pathspec 'gabi-kafka-client-zstd-codec' did not match any file(s) known to git
ERROR: Command errored out with exit status 1: git checkout -q gabi-kafka-client-zstd-codec Check the logs for full command output.
```

However, it seems the issue is I needed to prepend it with `git+`

```
-e git+git://github.com/aiven/kafka-python@gabi-kafka-client-zstd-codec#egg=kafka-python
```
^ works :)

**Comment by Rohit-Singh3** [comment]
**getting below error when trying to consume from a ZSTD compressed record from kafka.**

 **raise UnsupportedCodecError(
kafka.errors.UnsupportedCodecError: UnsupportedCodecError: Libraries for zstd compression codec not found**

**error log:**

Traceback (most recent call last):
  File "/home/PycharmProjects/myPythonLearning/kafka/consumer/consumer.py", line 25, in <module>
    consume_from_topic(topic_to_consume)
  File "/home/PycharmProjects/myPythonLearning/kafka/consumer/consumer.py", line 14, in consume_from_topic
    for message in consumer:
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/group.py", line 1193, in __next__
    return self.next_v2()
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/group.py", line 1201, in next_v2
    return next(self._iterator)
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/group.py", line 1116, in _message_generator_v2
    record_map = self.poll(timeout_ms=timeout_ms, update_offsets=False)
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/group.py", line 655, in poll
    records = self._poll_once(remaining, max_records, update_offsets=update_offsets)
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/group.py", line 708, in _poll_once
    records, _ = self._fetcher.fetched_records(max_records, update_offsets=update_offsets)
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/fetcher.py", line 344, in fetched_records
    self._next_partition_records = self._parse_fetched_data(completion)
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/fetcher.py", line 818, in _parse_fetched_data
    unpacked = list(self._unpack_message_set(tp, records))
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/consumer/fetcher.py", line 467, in _unpack_message_set
    for record in batch:
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/record/default_records.py", line 276, in __iter__
    self._maybe_uncompress()
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/record/default_records.py", line 183, in _maybe_uncompress
    self._assert_has_codec(compression_type)
  File "/home/PycharmProjects/myPythonLearning/venv/lib/python3.8/site-packages/kafka/record/default_records.py", line 118, in _assert_has_codec
    raise UnsupportedCodecError(
kafka.errors.UnsupportedCodecError: UnsupportedCodecError: Libraries for zstd compression codec not found

Process finished with exit code 1


**My Code for consuming a kafka topic:**

from  kafka  import KafkaConsumer
def consume_from_topic(topic):
    consumer = KafkaConsumer(
        topic,
        bootstrap_servers='localhost:9092', 
        group_id='zstd-11-consumer-group',   
        auto_offset_reset='earliest',       
        enable_auto_commit=True        
    )
    try:
        for message in consumer:
            v = message.value
            k = message.key.decode("utf-8")
            log = "key={}, offset={}, partition={}, value={}".format(k, message.offset, message.partition, v)
            print(log)

    except KeyboardInterrupt:
        consumer.close()
if __name__ == "__main__":
    topic_to_consume = "Integrate-Package-Zstd-ESP.info" 
    consume_from_topic(topic_to_consume)

**Comment by brokenjacobs** [comment]
Did you install the zstandard package from pypi?
